### ISSUE TITLE
Add a method to check intersection between a rect and a line

### ISSUE THREAD

**Issue by thiagojobson** [pygame/pygame#1406]
The [SDL_IntersectRectAndLine](https://wiki.libsdl.org/SDL_IntersectRectAndLine) can be used to check intersection beetwen a rect and a line.

This could be something like:
`Rect.collideline((x1, y1), (x2, y2)) -> bool`

**Comment by mcpalmer1980** [comment]
I would love that too. But collide_segment() probably shouldn't return a bool. Most of the time you would want to know which edge it collided with first. What we really need is some sort of raycasting functionality, which I have suggested before. https://github.com/pygame/pygame/issues/1152

**Comment by illume** [comment]
@thiagojobson Good idea about a `Rect.collideline`
- It matches the other collide* functions (rect, point, ...)
- should be able to reuse the SDL_IntersectRectAndLine code.

Should it take a line like `((x0, y0, x1, y1))` as well/instead?

We could start with a Bool (because it's easy to use SDL_IntersectRectAndLine), and move to returning an edge(or intersection?) at a further date?

**Comment by mcpalmer1980** [comment]
So I took a quick look at the docs for SDL_IntersectRectAndLine and it confirmed my assumption that it checks collisions for a segment, not a line. Also, although it returns a bool it also clips the line by modifying the line coordinate pointers. It then becomes almost trivial to determine which edge it collided with.

As for the python implementation I remain unsure. Other then with rects, I cannot think of any examples where pygame changes the values of a parameter. Perhaps collideline() should return the clipped line coordinates in the event of a collision, or false otherwise. Icing on the cake would order the returning endpoints so that the colliding edge is returned first so I don't need to perform a bunch of comparisons to determine where the collision occurred.

Then we need to consider how to describe our lines in a convenient way that fits with the other python methods. In pygame, 2-tuples seem the most common way to describe a point. I think that the collision method should be:
```
Rect.collideline((x1, y1), (x2, y2)) -> [(x1,y1), (x2,y2)] or False
```
We can then use it by calling something like:
``` bool
line = (enemy.get_pos(), target.get_pos())
clipped_line = my_sprite.rect.collideline(*line)
explosions.add(clipped_line[0])
```
As a side question, is pygame starting to adapt vectors to be used place of two-tuples? Should this be considered in the collidedline method? Because the pygame1.91 included in my distribution still lacked official vector support, I created my own class and have so much code that uses it that I never experimented with the vectors in pygame.

As for my earlier mention of raycasting, that doesn't belong in Rect anyway. It belongs either in a tilemap class or a physics engine. I have experimented with pymunk with pygame2 and like it, but unfortunately it doesn't seem to do raycasting :(

**Comment by illume** [comment]
Yeah, math.Vector* is considered stable at this point. Support was added to some functions already, and other ones already supported them.

**Comment by mcpalmer1980** [comment]
I don't see any method to return a Vector2 as a 2-tuple, which seems strange to me. Something like vector.tuple_float() and vector.tuple_int(). I guess it's not really necessary because it allows accessing x and y as if it was a tuple, something my own class doesn't do.

I really like my Vector class but I think I should switch over. I'm sure yours is faster, but I happen to prefer my naming conventions and CamelCase.
```
class Vector2:
	__init__(self, x=0, y=0): > Vector2
	FromPoints(self, P1, P2): > Vector2
	FromAngle(self,ang, mag=1): > Vector2
	GetAngle(self): > degrees
	GetMagnitude(self): > float
	SetMagnitude(self, mag): > Vector2
	Normalize(self): > Vector2
	Normal(self): > Vector2
	NormalMag(self): > Vector normal, int magnitude
	Shrink(self, amount): > float magnitude
	Tuple(self): ( x, y )
	TupleInt(self): int( x, y )
	__repr__(self):
	__str__(self):
	__add__(self, rhs):
	__mul__(self, scalar):
	__truediv__(self, scalar):
	__neg__(self):
	__sub__(self, rhs):
	__len__(self): > magnitude
	__nonzero__(self):
	__eq__(self, other):
	__ne__(self, other):
```

**Comment by illume** [comment]
`tuple(vec)` is ok for that I guess?

**Comment by thiagojobson** [comment]
I first thought it should return a bool so it would be consistent with the other collide functions. But as @mcpalmer1980 pointed out, it would be way more useful to just return the clipped line as SDL already does the calculation.

> `Rect.collideline((x1, y1), (x2, y2)) -> [(x1,y1), (x2,y2)] or False`

I like this because the return value can still be evaluated to a bool. But pygame functions historically accept parameters with different formats, so `((x1, y1, x2, y2))` is to consider.

**Comment by mcpalmer1980** [comment]
> tuple(vec) is ok for that I guess?

Yes, in most situations. I starting writing that post before I realized that you can access a Vector as a tuple, which I never thought about when I created my Vector class. I didn't really know python back then(15mo ago). For instance, my menu class uses ordered parameters and is terribly error prone because I remained unaware of dict parameters.

 I have the Vector.TupleInt() method for 2 reasons: one, accessing a grid array requires int and to make comparisons with grids easier. Using pygame's vector class could get a bit clumbsy in these situations.

grid[int(vec.x), int(vect.y)]
if int(vec.x) == this.x and int(vec.y) == this.y:
etc.

**Comment by mcpalmer1980** [comment]
I just thought of something. This functions will solve a long standing issue I've had with my games. In order for it to work collideline must return the intersect point, which will be one of the clipped line's endpoints, or at least return the intersecting endpoint as the first point of the line as I suggested earlier.

**Example: player.rect is colliding with box.rect**
How do I know if I should make the player stand on the box or push the box? My method used the rect.clip() method but was inaccurate when the corners of both rects collided. With clipline it would be better.

```
for b in player.rect.collidelistall(boxes):
	box = boxes[b]
	point = box.rect.collideline(player.center)
	if point.x == box.left:
        	player.touch_right = box
	elif point.x == box.right:
		player.touch_left = box
	elif point.y == box.top:
		player.standing_on = box
	elif point.y == box.bottom:
		player.bump_head = box magically floating
	else:
		I don't know what just happened
```

**Comment by charlesej** [comment]
I have developed code for this, but would like some agreement on the API before creating a PR. My current code has the following API.

> collideline((x1, y1, x2, y2)) -> ((x1, y1), (x2, y2))
> collideline((x1, y1, x2, y2)) -> ()
> collideline(((x1, y1), (x2, y2))) -> ((x1, y1), (x2, y2))
> collideline(((x1, y1), (x2, y2))) -> ()

- returns a `tuple` (instead of a `list`)
- returns an empty tuple (instead of `False`) when there is no collision
- takes the following arguments
  - `tuple`/`list` of 4 ints: e.g. `collideline((x1, y1, x2, y2))`
  - `tuple`/`list` of 2 `tuple`/`list`/`Vector2` of 2 ints: e.g. `collideline(((x1, y1), (x2, y2)))`

With the current API:
- The returned value can easily be tested to see if a collision occurred
- If a collision has occurred:
  - The returned value (line) can be easily drawn.
  - The returned value (line) of a `collideline()` call can be passed to another `collideline()` call.

Here's some example code showing the usage of `collideline()`.
```python
import pygame

pygame.init()
screen = pygame.display.set_mode((300, 400))
rect = pygame.Rect((10, 20), (30, 50))
pygame.draw.rect(screen, pygame.Color('blue'), rect)

for line in (((1, 25), (200, 25)), ((1, 90), (200, 90))):
    collideline = rect.collideline(line)
    print('collideline =', collideline)

    pygame.draw.line(screen, pygame.Color('green'), line[0], line[1])

    if collideline:
        pygame.draw.line(screen, pygame.Color('red'), collideline[0],
                         collideline[1])

pygame.display.flip()
pygame.quit()
```

Here's the text output of the above code.
```
pygame 2.0.0.dev5 (SDL 2.0.10, python 3.7.4)
Hello from the pygame community. https://www.pygame.org/contribute.html
collideline = ((10, 25), (39, 25))
collideline = ()
```

Should the current API be changed or added to? Other formats to consider:
- 4 ints: e.g. `collideline(x1, y1, x2, y2)`
- 2 `tuple`/`list`/`Vector2` of 2 ints: e.g. `collideline((x1, y1), (x2, y2))`

**Comment by mcpalmer1980** [comment]
I actually think the two recommendations at the end of your post are the most common use cases. 

**Best**
+collideline(x1, y1, x2, y2)
good basic usage.

+collideline((x1,y1), (x2, y2))
supports Vector2s and common 2-tuple points

**Useful**
collideline(((x1, y1), (x2, y2)))
Having a single 2-tuple pair like this seems unlikely, but looping through a long list of 2-tuple lines might be common, and this would be the best way to pass in each line. We could do without it, however, like this.
```
for line in lines:
    collide = rect.collideline(*line)
```

**Unnecessary**
collideline( (x1,y1, x2, y2) )
Rects might be stored like this, but I doubt lines will be. They will be a series of two points. If you do have a 4-tuple you want to send to collide line, collideline( *line ) would work anyway.

**Comment by mcpalmer1980** [comment]
There are two other considerations.

The rect.collideline() method must not change the parameter values as the SDL_IntersectRectAndLine() function does. Of course I am only just starting to look at pygame's c source and your activity in this repository suggests you have cardinal knowledge, so you must have already considered this if it bares consideration.

My request to return the colliding point first should not be overlooked. Most people calling this method will want to know WHERE the line collided.

Sure we could do a few comparisons to discern this if you return the points in the order they were received, but it would be quicker in C and lead to MUCH cleaner python code. Also, I cannot think of any situations where changing the order of the points should have any adverse effect.

**Comment by charlesej** [comment]
@mcpalmer1980 thanks for the comments. I will look at adding support for the other 2 formats.

Relating to your other considerations:

> The rect.collideline() method must not change the parameter values as the SDL_IntersectRectAndLine() function does.

- The parameter values won't be changed.

> My request to return the colliding point first should not be overlooked. Most people calling this method will want to know WHERE the line collided.

- Is this a correct interpretation of your request?
  - If a collision occurs between a line and a rect, the returned line is clipped to the rect. The order of the returned line endpoints should be as follows:
    - If one end of the line has been clipped and the other has not, the clipped end should be first in the line sequence.
    - If both ends are clipped or if neither are clipped then the original parameter point order is fine.

**Comment by mcpalmer1980** [comment]
I see 4 possible situations. You got the 2 common ones right, but I'm not sure how the other 2 could be handled. It would be easy to handle those in a basic BOOL collision test, but not as a clipped line, which would be more useful. 

* one point is inside rect - return the clipped endpoint first.
* neither point is inside rect - return () or False I think, because there has been no collision.
* both points are inside rect - I don' t know how SDL handles that. Does it extend the line segment and return the points where that segment collide with the rect? If so, return in the original order.
* both points are outside rect but it passes through: I don't know how SDL handles that, clipping would create 2 line segments.

Perhaps you should just ignore my suggestion and return the values in the original order. I overlooked something that should have been obvious. My earlier use example looked like this.
```
line = (player.get_pos(), target.get_pos())
clipped_line = my_sprite.rect.collideline(*line)
if clipped_line:
    explosions.add(clipped_line[0])
```
I already know that the first point is the player, so the second point is the one that would be clipped. That's the point where I would draw an explosion, for example. There are times when someone might be working with arbitrary lines and want to know which side hit, but I'm not sure how often that would happen.

**Comment by charlesej** [comment]
@mcpalmer1980 I will keep the order of the returned points the same as they are passed in.

**Comment by charlesej** [comment]
Any objections to changing the name of the method to `clipline()` (that is what it is actually doing)?

Here's the API that will be supported:
> clipline((x1, y1), (x2, y2)) -> ((x1, y1), (x2, y2))
> clipline((x1, y1), (x2, y2)) -> ()
> clipline(x1, y1, x2, y2) -> ((x1, y1), (x2, y2))
> clipline(x1, y1, x2, y2) -> ()
> clipline(((x1, y1), (x2, y2))) -> ((x1, y1), (x2, y2))
> clipline(((x1, y1), (x2, y2))) -> ()
> clipline((x1, y1, x2, y2)) -> ((x1, y1), (x2, y2))
> clipline((x1, y1, x2, y2)) -> ()

**Comment by mcpalmer1980** [comment]
Keeping the order seems fine upon further reflection. Don't know why it seemed so important to me before. I think you're going too far with the methods of input, but if it doesn't effect performance I'm sure someone will appreciate them all at some point.

clipline() does seem a more appropriate title and fits in with the rect.clip() naming. I see no reason to stick with the SDL name since it doesn't really make sense.

I wonder if you experimented with colliding a line that's completely inside a rect. What sort of results would it return, and importantly, will it evaluate to true? It is possible that the SDL_IntersectRectAndLine() function returns True in this case but sets the int refs to 0 since it the line never actually touches the edge. The SDL docs aren't clear about this.

**Comment by charlesej** [comment]
As per the Remarks section of [SDL_IntersectRectAndLine](https://wiki.libsdl.org/SDL_IntersectRectAndLine): "A line segment contained entirely within the rectangle or that does not intersect will remain unchanged."

> Remarks
This function is used to clip a line segment to a rectangle. A line segment contained entirely within the rectangle or that does not intersect will remain unchanged. A line segment that crosses the rectangle at either or both ends will be clipped to the boundary of the rectangle and the new coordinates saved in X1, Y1, X2, and/or Y2 as necessary.

**Comment by thiagojobson** [comment]
Thanks for working on this request @charlesej.

+1 to `clipline` as it is the appropriated name. I really like how the method signature is designed and that it returns either a clipped line or an `()` (empty line?) when there's no overlapping. 

I think the input parameters:

> clipline((x1, y1), (x2, y2)) -> ...
clipline(x1, y1, x2, y2) -> ...

should be preferred over:

> clipline(((x1, y1), (x2, y2))) -> ...
clipline((x1, y1, x2, y2)) -> ...

as they are special cases that can be accomplished if we unpack the parameters when calling the method.

```python
line1 = ((x1, y1), (x2, y2))
line2 = (x1, y1, x2, y2)

clipline(*line1)
clipline(*line2)
```

**Comment by mcpalmer1980** [comment]
For our purposes clipline() is an appropriate title because it does clip the line. But it was called collideline because it returns a bool if the the line touches. This includes situations where the line fits entirely within the rect.

Keep in mind that the original request was for a method to determine if a line collides with a rect. If it does not touch the rect, or if it is contained within the rect, we do not want to receive the original rect's endpoints. We want a value that resolves to False.

In this respect, the name collideline does make sense.

I agree with thiagojobson's assessment of parameter options.

**Comment by charlesej** [comment]
@thiagojobson thanks for the feedback.

**Comment by charlesej** [comment]
@illume what are your thoughts on the API?

**Comment by charlesej** [comment]
@mcpalmer1980 if a line is fully inside a rect that would be considering colliding with it. The term colliding in this context means overlapping.

Just like the other rect collision methods.

- `collidepoint()` - any point inside a rect is considered to collide with it
  > Returns true if the given point is inside the rectangle. A point along the right or bottom edge is not considered to be inside the rectangle.

- `colliderect()` - any rect fully inside another rect is considering colliding with it
  > Returns true if any portion of either rectangle overlap (except the top+bottom or left+right edges).

If you are looking for a way to see if a line crosses a rect's border, the `clipline()` method can be used to provide that functionality. If the returned line is different than the passed in line, then it has been clipped, which means that it crossed the rect's border. (These following code snippets assume the format of `line` being passed into `clipline()` is `((x1, y1), (x2, y2))`, otherwise the code will need to be modified accordingly.)
```python
clippedline = rect.clipline(line)

if clippedline:
    # line collides/overlaps with rect

    if clippedline == line:
        # line fully inside rect
    else:
        # line crosses the rect's border
else:
    # line fully outside rect
```

Here's a simple function that returns a clipped line only if the original line crosses the rect's border.
```python
def rect_border_crossing_clipline(rect, line):
    clippedline = rect.clipline(line)
    return clippedline if clippedline and clippedline != line else ()
```

**Comment by illume** [comment]
TLDR; I think `clipline` with docs changes near colliderect, and collidepoint to say clipline can be used for a collideline will be ok. A little example of how to see where the collision starts would be nice (see bottom of post for example). Args are good imho.


I wonder if people will realise rect.clipline can be used as a collideline. Maybe if we write in the docs near collidepoint, and colliderect that clipline can be used as a collideline. 

The following code is sort of readable...
```python
if alien.rect.clipline(line):
    Explosion(alien)
```

But maybe not as readable as this:
```python
if alien.rect.collideline(line):
    Explosion(alien)
```

Is introducing an alias is a good idea?
(`Rect.clipline`, AND `Rect.collideline`).

I'm not sure.


One other thing is the return value is not clear to me from just reading the code.
An example like this shows the structure of the return values.
```python

# simple example of collision detection
if alien.rect.clipline(line):
    alien.kill()

# see where the line clips with the rect, and explode there.
collision = alien.rect.clipline(line):
if collision:
    start, end = collision
    x, y = start
    Explosion((x, y))
    alien.kill()
```

pygame.draw.line takes a start_pos, and end_pos, and draw.lines takes a list of points.

**Comment by mcpalmer1980** [comment]
**@ charlesej**
I apologize for any confusion. I was worried about the possibility of a line being returned where a false value should be returned, but somehow the logic got reversed in my head and I explained it all wrong.

**Comment by charlesej** [comment]
@illume thanks for your input.

**Comment by mcpalmer1980** [comment]
@charlesej You mentioned the rect.right and rect.bottom 'missing pixel' in your PR. Have you determined whether the clipped line's endpoint will exactly match the right and bottom attributes for of the rect for comparisons?

**Comment by charlesej** [comment]
The intention of this note is to remind users how rect values work.

> Note: The rect.bottom and rect.right attributes of a pygame.Rectpygame object for storing rectangular coordinates always lie one pixel outside of its actual border.

At the top of the rect docs there is also this.

> The area covered by a Rect does not include the right- and bottom-most edge of pixels. If one Rect's bottom border is another Rect's top border (i.e., rect1.bottom=rect2.top), the two meet exactly on the screen but do not overlap, and rect1.colliderect(rect2) returns false.

---

Here's some code to help explain how this works in relation to `clipline()`. As shown below, the values returned by `rect.topright`, `rect.bottomleft`, and `rect.bottomright` are not actually inside the rect. If a line is made from `rect.topleft` to `rect.topright` and passed to `clipline()`, the line's topright endpoint will be clipped by 1 pixel. As mentioned above, this is true for all the `rect.bottom` and `rect.right` attributes. So in the example below, a line of `(rect.topright, rect.bottomright)` or `(rect.bottomleft, rect.bottomright)` will not collide with the rect.

```python
import pygame

rect = pygame.Rect((0, 0), (30, 30))
print('rect =', rect)
print('\nrect.collidepoint(rect.topleft) =', rect.collidepoint(rect.topleft))
print('rect.collidepoint(rect.topright) =', rect.collidepoint(rect.topright))
print('rect.collidepoint(rect.bottomleft) =', rect.collidepoint(rect.bottomleft))
print('rect.collidepoint(rect.bottomright) =', rect.collidepoint(rect.bottomright))


line = (rect.topleft, rect.topright)
print('\nline =', line)
print('rect.clipline(line) =', rect.clipline(line))

line = (rect.topleft, rect.bottomleft)
print('\nline =', line)
print('rect.clipline(line) =', rect.clipline(line))

line = (rect.topleft, rect.bottomright)
print('\nline =', line)
print('rect.clipline(line) =', rect.clipline(line))

line = (rect.topright, rect.bottomright)
print('\nline =', line)
print('rect.clipline(line) =', rect.clipline(line))

line = (rect.bottomleft, rect.bottomright)
print('\nline =', line)
print('rect.clipline(line) =', rect.clipline(line))
```

```
pygame 2.0.0.dev5 (SDL 2.0.10, python 3.7.4)
Hello from the pygame community. https://www.pygame.org/contribute.html
rect = <rect(0, 0, 30, 30)>

rect.collidepoint(rect.topleft) = 1
rect.collidepoint(rect.topright) = 0
rect.collidepoint(rect.bottomleft) = 0
rect.collidepoint(rect.bottomright) = 0

line = ((0, 0), (30, 0))
rect.clipline(line) = ((0, 0), (29, 0))

line = ((0, 0), (0, 30))
rect.clipline(line) = ((0, 0), (0, 29))

line = ((0, 0), (30, 30))
rect.clipline(line) = ((0, 0), (29, 29))

line = ((30, 0), (30, 30))
rect.clipline(line) = ()

line = ((0, 30), (30, 30))
rect.clipline(line) = ()
```
